In [2]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict
from sklearn.ensemble import RandomForestClassifier

In [5]:
data=pd.read_csv(r"C:\Users\comp\Downloads\heart.csv")
data

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[918 rows x 12 columns]

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


# LabelEncoding

In [4]:
from sklearn.preprocessing import LabelEncoders
L=LabelEncoder()
data["Sex"]=L.fit_transform(data["Sex"])
data["ChestPainType"]=L.fit_transform(data["ChestPainType"])
data["RestingECG"]=L.fit_transform(data["RestingECG"])
data["ExerciseAngina"]=L.fit_transform(data["ExerciseAngina"])
data["ST_Slope"]=L.fit_transform(data["ST_Slope"])
data

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0     40    1              1        140          289          0           1   
1     49    0              2        160          180          0           1   
2     37    1              1        130          283          0           2   
3     48    0              0        138          214          0           1   
4     54    1              2        150          195          0           1   
..   ...  ...            ...        ...          ...        ...         ...   
913   45    1              3        110          264          0           1   
914   68    1              0        144          193          1           1   
915   57    1              0        130          131          0           1   
916   57    0              1        130          236          0           0   
917   38    1              2        138          175          0           1   

     MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0      172               0      0.0         2             0  
1      156               0      1.0         1             1  
2       98               0      0.0         2             0  
3      108               1      1.5         1             1  
4      122               0      0.0         2             0  
..     ...             ...      ...       ...           ...  
913    132               0      1.2         1             1  
914    141               0      3.4         1             1  
915    115               1      1.2         1             1  
916    174               0      0.0         1             1  
917    173               0      0.0         2             0  

[918 rows x 12 columns]

In [5]:
data["HeartDisease"].value_counts()

1    508
0    410
Name: HeartDisease, dtype: int64

# data slicing

In [6]:
x=data.iloc[:,:-1].values
y=data.iloc[:,-1].values

# Imbalancing method SMOTE

In [7]:
from imblearn.over_sampling import SMOTE
S=SMOTE()
x_data,y_data=S.fit_resample(x,y)

# Featuer scaling

In [8]:
from sklearn.preprocessing import StandardScaler
SS=StandardScaler()
X_data=SS.fit_transform(x_data)

In [9]:
X_data

array([[-1.4320549 ,  0.54383702,  0.20066802, ..., -0.78817095,
        -0.81731928,  0.99958847],
       [-0.46337404, -1.92130733,  1.26315768, ..., -0.78817095,
         0.14856614, -0.6635692 ],
       [-1.75494852,  0.54383702,  0.20066802, ..., -0.78817095,
        -0.81731928,  0.99958847],
       ...,
       [ 0.69099398,  0.20514692, -0.86182163, ..., -0.78817095,
        -0.4877423 ,  0.77108459],
       [-1.49480097,  0.54383702,  0.20066802, ..., -0.78817095,
         0.71165116,  0.99958847],
       [-0.07170698, -1.13372684,  0.54011918, ..., -0.78817095,
        -0.81731928,  0.99958847]])

# Splinting the data in train and testing sets

In [10]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(X_data,y_data,test_size=0.2,random_state=8)

In [11]:
from collections import Counter
Counter(y_test)

Counter({1: 107, 0: 97})

# Using logistic regression

In [12]:
from sklearn.linear_model import LogisticRegression
L=LogisticRegression()

In [13]:
L.fit(x_train,y_train)

LogisticRegression()

In [14]:
yl=L.predict(x_test)
yl

array([1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0,
       0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0,
       1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1,
       1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0,
       1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0,
       0, 0, 1, 1, 1, 0], dtype=int64)

In [34]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,yl)*100

84.80392156862744

# The accuracy_score of the Logistic Regression is the 84.80 %

# cross validation method kfold

In [16]:
from sklearn.model_selection import KFold
kf=KFold(n_splits=5,random_state=16,shuffle=True)
kf.get_n_splits(X_data)
print(kf)

KFold(n_splits=5, random_state=16, shuffle=True)


In [17]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict

In [18]:
score=cross_val_score(L,x_train,y_train,cv=kf)
y_pred=cross_val_predict(L,x_test,y_test,cv=kf)
print(score)
print("mean of the all folds",np.mean(score)*100)
print(y_pred)

[0.85276074 0.81595092 0.85185185 0.82716049 0.87654321]
mean of the all folds 84.48534423994548
[1 0 1 0 1 1 1 0 1 1 1 1 0 1 1 1 1 1 0 0 1 1 0 0 0 1 1 1 0 1 0 1 1 1 1 0 1
 0 1 1 0 0 1 1 1 0 0 1 0 0 0 1 0 0 1 0 1 1 0 1 0 1 1 0 0 0 0 0 1 1 1 0 1 0
 1 1 0 0 0 1 0 1 1 1 1 1 1 0 1 1 1 1 1 1 0 1 0 1 0 1 0 0 0 0 0 1 1 0 0 0 1
 0 1 1 1 1 0 0 0 0 0 1 0 0 1 0 0 1 0 0 0 1 1 1 0 1 1 0 1 1 1 0 0 1 1 0 0 1
 0 1 0 0 1 1 1 0 0 0 0 1 1 1 0 0 1 1 1 1 0 1 1 1 1 0 0 1 1 1 1 1 0 1 0 0 0
 0 1 0 1 1 0 1 0 1 1 1 0 0 0 0 1 1 1 0]


In [19]:
accuracy_score(y_test,y_pred)*100

84.80392156862744

# After using the kfold the accuracy score is 84.80 %

# Ensemble Techniques

In [21]:
from sklearn.ensemble import RandomForestClassifier
R=RandomForestClassifier()
R.fit(x_train,y_train)

RandomForestClassifier()

In [23]:
yr=R.predict(x_test)
yr

array([1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0,
       0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0,
       1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0,
       0, 0, 0, 1, 1, 0], dtype=int64)

In [25]:
accuracy_score(y_test,yr)*100

88.72549019607843

# the accuracy score of the Random forest Classifier is 88.72%

In [29]:
from sklearn.neighbors import KNeighborsClassifier
k1=KNeighborsClassifier(n_neighbors=5,metric="minkowski",p=1)
model=k1.fit(x_train,y_train)
model

KNeighborsClassifier(p=1)

In [31]:
from sklearn.ensemble import VotingClassifier 
max_vote=VotingClassifier(estimators=[('random Forest',R),("Logistic Regression",L),("KNN:",k1)])
max_vote

VotingClassifier(estimators=[('random Forest', RandomForestClassifier()),
                             ('Logistic Regression', LogisticRegression()),
                             ('KNN:', KNeighborsClassifier(p=1))])

In [32]:
model=max_vote.fit(x_train,y_train)
model

VotingClassifier(estimators=[('random Forest', RandomForestClassifier()),
                             ('Logistic Regression', LogisticRegression()),
                             ('KNN:', KNeighborsClassifier(p=1))])

In [38]:
M=model.predict(x_test)
M

array([1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0,
       0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0,
       1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0,
       0, 0, 1, 1, 1, 0], dtype=int64)

In [39]:
aa=accuracy_score(y_test,M)*100
print("VotingClassifier accuracy",aa)



VotingClassifier accuracy 87.25490196078431
